### ISSUE TITLE
Tracks folder listing is mixed up (Configure map>tracks>menu)

### ISSUE THREAD

**Issue by Rowin63** [osmandapp/OsmAnd#19634]
### Description

The track folders are not listed aphabetical anymore, but mixed up

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/95540936/8f1124d7-681d-4136-91da-dd7b75c44f59"  width="300" height="auto"/> </td>
  </tr>
</table>

### Steps to reproduce

go to "configure map / tracks / folder menue"

### Actual result

after changing some folders names the track folders are not in alphabetical order anymore

(but in "my places / tracks" the listing is correct)

### Expected result

keep them in order

### Your Environment (required)

OsmAnd Version: +4.7.10
Android/iOS version: 13
Device model: S20+

**Comment by Rowin63** [comment]
Sotty, double!

**Comment by sonora** [comment]
It's a regressiion, and seems separate from and in addition to #19333.

**Comment by sonora** [comment]
Apparently, the function "Sort subfolders" on the MyTracks screen also successfully sorts all **tracks** of 'folder' "Visible on map" and on its associated screen, but not the folder list itself on that screen (neither the folder tabs nor the folder list).

**Comment by sonora** [comment]
The folder list on the 'Visible on map' screen and list menu seems always by the folders' time stamp, I find no way currently to sort it alphabetically.

**Comment by vshcherb** [comment]
To fix: The order on Configure Map should be identical to My places screen

**Comment by sonora** [comment]
It looks like now we 'reversed' the issue: The function "Sort subfolders" on the MyTracks screen now successfully also sorts all folders on "Visible on map", but now all tracks in the folders on "Visible on map" are always sorted by 'Last modified".

**Comment by vshcherb** [comment]
- Reverted fix - https://github.com/osmandapp/OsmAnd/pull/19682
- And switched to 4.8
Main issues:
1. Too much code (it looks suspicous that we implement different sorting algorithms) on My Places and here
2. Didn't fix the issue ? 
3. Introduced delay - non-optimal

**Technical note**
- We should have simple API TracksManager.getFolderList() - which will return list of tabs in order (saved in My Places)
- That API should be used in Configure Map and In My Places

**Comment by sonora** [comment]
@Chumva  I think the issue may be the sort mode for the "root" folder in the MyPlaces>Tracks screen: That very sort mode should be also used
-  **for the "ALL"** tab of the Configure map>Tracks folder,
- as well as for its **tabs**
- and its **tab dialog** order.

Perhaps in https://github.com/osmandapp/OsmAnd/blob/21d390dc18ea61bbfaf60f97288f439e7d6015eb/OsmAnd/src/net/osmand/plus/configmap/tracks/TrackTabsHelper.java#L314 that sort mode receives a wrong key in the Map which is then not used for the 3 purposes??

**Comment by sonora** [comment]
From what I see, if we merge https://github.com/osmandapp/OsmAnd/pull/19814, we now just need to add code to sort 'Configrue map>Tracks' tabs according to the GPX-root-folde's sortMode as set in the 'MyPlaces>Tracks' screen, and spresent in `TRACKS_TABS_SORT_MODES`.

**Comment by Chumva** [comment]
Yes, also we need to make it a universal code, for my places & configure tracks

**Comment by sonora** [comment]
@Chumva  here's a hotfix to resolve the current randomness: https://github.com/osmandapp/OsmAnd/commit/a6e896f814343e7577ed139e915a14502197d53a. We can refine the code as we go along, ok?

**Comment by sonora** [comment]
PS: Perhaps we can even consider this issue solved for now, I am not sure it makes much sense in practice to offer more tab orders than A-Z and Z-A.

The hotfix uses the root folder's sorting also for the tab order, but is currently limited to A-Z and Z-A, defaulting to A-Z for all other root folder sort modes.

**Comment by vshcherb** [comment]
Technical debt left

**Comment by alex-osm** [comment]
Should be fixed already

**Comment by Rowin63** [comment]
~4.9.0#3431m
Folders still are not in correct alphabetical order in "configure map > tracks > folder list" 
(but they are in "my places > tracks")

![Screenshot_20241014_204745_OsmAnd~](https://github.com/user-attachments/assets/10f46e59-2d06-4ab4-b187-d43a22fa2327)

**Comment by Rowin63** [comment]
Just as an additional info:
Recently I´ve created some "main" folders including "sub" folders.
The listing of these folders also doesn´t seem to be correct, they´re listed by the "sub" folders name:

![Screenshot_20241112_070139_OsmAnd+](https://github.com/user-attachments/assets/3916b504-55b3-4cf7-97f1-9f627664d4ed)

![Screenshot_20241112_070207_OsmAnd+](https://github.com/user-attachments/assets/392f260b-8142-4cd2-91fd-d3e44d60c858)

**Comment by sonora** [comment]
Still not fixed is the following (on the "_Visible on map_" screen, and on its "_Switch folder_" button ): Subfolders in A-Z mode appear listed according to their standalone name, but preferably they should be sorted after their parent folder:

**As-is sorting** (A, B, C folders; X, Y subfolders:):
```
A
B
C
A/X
B/Y
```

Better would be:
```
A
A/X
B
B/Y
C
```

**Comment by sonora** [comment]
What's worse: Now the sort order is not preserved between app starts (I have things A-Z, it always jumps back to "Last modified".)

**Comment by Zirochkabila** [comment]
`OsmAnd~ 5.0.0#3836m, Released: 2024-11-28`
Sort subfolders does not work

https://github.com/user-attachments/assets/cec049e6-9a59-4546-9fbe-ab21902f5d12

**Comment by Rowin63** [comment]
> What's worse: Now the sort order is not preserved between app starts (I have things A-Z, it always jumps back to "Last modified".)

"3dots > sorting subfolders" still shows up as set to "A-Z" (it's "blue", selected), but actual sorting is "last modified"

**Comment by vshcherb** [comment]
Minor issue, we shouldn't highlight item

**Comment by sonora** [comment]
Yes, the highlighting  in the "Switch folders" list is a mere "suggestion", not the current value.

- [ ] But that's not the issue I am reporting, it is this: Sort the main "My Places / Tracks screen" itself A-Z, restart the app: It will fall back to "Last modified".
- [ ]  Second, but comparatively minor issue is the subfolders not being grouped under parent folders, see above.

**Comment by Rowin63** [comment]
> Minor issue, we shouldn't highlight item

Sorry, poorly expressed. 
What I wanted to say: "A-Z" was my selected option. After app restart it still is selected, but the sorting always falls back to "last edited".

**Comment by vshcherb** [comment]
To fix:

- [ ] But that's not the issue I am reporting, it is this: Sort the main "My Places / Tracks screen" itself A-Z, restart the app: It will fall back to "Last modified".
- [ ]  Second, but comparatively minor issue is the subfolders not being grouped under parent folders, see above.

**Comment by nazar-kutz** [comment]
# How sorting works currently
I decided to describe in a little more detail the principles of track sorting at the moment, to clarify certain points for this feature testing, support and further development.

### My Places -> Tracks
On the **My Places -> Tracks** screen, you can configure the sorting order separately for each standard folder (a physical directory on the device) and each smart folder (a set of customizable filters that helps structure tracks according to certain parameters).
* The selected sorting mode applies only to the content directly located in the folder.
* Content in subfolders does not inherit this sorting mode; instead, sorting for each subfolder is configured manually and independently.

To simplify the process of setting sorting modes for subfolders, you can use the **Sort subfolders** feature:
1. Open the folder whose subfolders need a sorting mode applied.
2. Tap the additional options button (**three-dots button**) in the toolbar at the top of the screen.
3. Select the "**Sort subfolders**" option from the dropdown menu.
4. In the dialog that appears, choose the desired sorting order.

**Key details:**
* The selected sorting order will be applied to all subfolders of the folder.
* The sorting mode of the folder itself (from which the option was triggered) will remain unchanged.
* Later, you can individually adjust the sorting mode for each subfolder as needed.
This feature simplifies the process of manually setting sorting modes, preventing the need to open each folder and its subfolders individually to configure sorting.
* When opened, this menu highlights the currently selected sort type for the folder. It is important to note that this menu does not display any information about the sort type selected for subfolders, as the content in them can be sorted in different modes. This highlighting can be seen  as a hint as to which mode the base folder is sorted in, if we want to set the same sort mode for subfolders.

**Smart folders features related to sorting**
Smart Folders have specific features:
* All Smart Folders have only **one level of nesting**, meaning they do not contain subfolders.
* Smart Folders are displayed directly in the root directory "**Tracks**" on the **My Places -> Tracks** screen, alongside regular first-level subfolders. For Android, this corresponds to the "Tracks" folder in the file system, and for iOS -- the "GPX" folder.

**Sorting Smart Folders**
* When changing the sorting order of the root folder "**Tracks**", regular folders and Smart Folders are sorted together as a **single list**.
* This results in both types being displayed intermixed, without separation into distinct categories.

### Configure Map -> Tracks
The **"Configure Map" -> "Tracks"** screen has a unique sorting approach tailored for managing the visibility of tracks and folders on the map.

**Key Features:**
1. **Tab structure instead of a tree view:**
   * Tabs are displayed as a horizontal list at the top of the screen.
   * Tab display order:
      1. **On map** -- shows all tracks currently visible on the map.
      2. **All** -- displays all available tracks.
      3. **Smart folders** -- sorted according to the sort mode set for the root "Tracks" folder via **My Places -> Tracks**
      4. **Physical folders** -- listed below smart folders, sorted using the same mode.
2. **Flattened folder structure:**
  * All folders are displayed as a single list without hierarchy showing only folder names.
 
It is also possible to view tabs as a dialog:
* The "**switch folder**" button (to the left of the "**three dots menu**"" button) opens a bottom sheet dialog.
* Unlike the horizontal menu, this dialog includes **parent folder name** for nested folders that placed deeper than in the root folder "Tracks".

**Tab Sorting Synchronization Features**
It is important to note that the order of tabs on the **"Configure Map" -> "Tracks"** screen is synchronized with the sorting mode selected for the **"Tracks"** root folder. It meens that every time the sorting mode for the "Tracks" tab is changed, the tabs order on this screen and in the "**Switch folder**" dialog updates instantly, reflecting the newly selected order.

**Comment by nazar-kutz** [comment]
# Further Tasks
During working on fixes for this issue and related, several another issues were discovered that are not critical for the current task but require separate consideration and, potentially, fixes within other tasks. These notes are documented here to avoid losing them:

### Saving sort mode in "Selection mode" (My Places -> Tracks)
On the `My Places` -> `Tracks` screen, it's possible to enter "**Selection mode**" (via `three-dots button` -> `Select` or by long-pressing any folder/track).
In this mode, a sort mode can be set for a folder, but it is not saved after exiting the mode.

### Temporary sorting on the track selection screen
On the track selection screen (e.g., when choosing "**Follow Track**"), it is possible to select a sorting mode for specific folders.
However, the selected sorting mode is temporary: after leaving the track selection screen, the sorting mode is reset.

### Duplicate folder names in "Configure Map -> Tracks"
Due to the possibility of overlapping names for subfolders and their parent folders, duplicate items may appear in the top tab menu and in the "Switch folder" dialog.
For example: for folders with relative paths "tracks/1/A/B" and "tracks/2/A/B", the top tab menu will display two tabs named "A" and two tabs named "B". In the "Switch folder" dialog, two entries will appear as "A/B".

Potential Solution:
* In the "Switch folder" dialog, consider using the full relative path to the folder instead of only including the parent folder's name.
* Keep the current approach in the top tab menu unchanged, as extending folder names there might cause usability issues for users.

**Comment by sonora** [comment]
@nazar-kutz , @Chumva , @vshcherb Thanks fot the exlanation and actions! But something is still not right let me summarize once again and demonstrate with screenshots. I have highlighted the issues in yellow.

**_I have set "A-Z" for all folders and all subfolders on the My Tracks screen!_**

**Case 1**, actual folder structure:
```
A
A/X
B
B/Y
C
import
rec
```
results in
<table style="width:100%">
  </tr>
  <tr>
    <td> <image source src="https://github.com/user-attachments/assets/60f309b2-c6bb-46f5-ae19-a6dc49a53a42"  width="300" height="auto"/>   </td>
    <td><image source src="https://github.com/user-attachments/assets/4d20af31-0996-40df-8951-8c6dfcdedd25"  width="300" height="auto"/>   </td>
  </tr>
</table>

**Case 2:**, actual folder structure:
```
import
myTracks    (contains no tracks, only subfolders)
myTracks/A
myTracks/A/X
myTracks/B
myTracks/B/Y
myTracks/C
rec
```
results in:
<table style="width:100%">
  <tr>
    <td> <image source src="https://github.com/user-attachments/assets/bd38eb3a-7b4b-4aa5-ba65-adb9b6a65daa"  width="300" height="auto"/>   </td>
    <td><image source src="https://github.com/user-attachments/assets/f97c0aa6-51bb-4638-972e-a6874c7d8331"  width="300" height="auto"/>   </td>
  </tr>
</table>

**Comment by nazar-kutz** [comment]
@sonora, thank you for your comments.

It appears that the first case is functioning as intended, since the “On Map” and “All” tabs on the Configure Map -> Tracks screen operate with their own sorting modes. In other words, their sort settings are independent from the sort mode selected in My Places -> Tracks.

Regarding the second case, I agree that it would be better to display the full relative path, at least in the "Switch folder" dialog. This issue is also mentioned [here](https://github.com/osmandapp/OsmAnd/issues/19634#issuecomment-2514900649) under the heading "Duplicate folder names in Configure Map -> Tracks."

**Comment by sonora** [comment]
@nazar-kutz Let's re-open this to fix
- [ ] _... I agree that it would be better to display the full relative path, at least in the "Switch folder" dialog._ 

ok? I seen no reason why we would not use **the full sub-path from the tracks root folder** for the "A-Z" sorting, I guess everything else produces weird effects.

**Comment by vshcherb** [comment]
Also it's hard to follow already this issue

**Comment by Rowin63** [comment]
When moving a track (my places > any track folder > 3dots of the track > move) the appearing folder list still is mixed up ("tracks" is the first folder in this list, "imported" and "rec" are the last ones all the way down) 

Osmand +4.9.10, Android 13
